In [20]:
""" Final Project """

import random
import math
import matplotlib.pyplot as plt
import pandas as pd
import os
from sklearn.feature_extraction.text import TfidfVectorizer 
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import HashingVectorizer
from sklearn.svm import SVC
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import KFold
from sklearn.linear_model import LinearRegression
import numpy as np

In [2]:
from sklearn.feature_extraction.text  import TfidfVectorizer
from sklearn.svm import SVC
from sklearn.metrics import classification_report, precision_score, recall_score

In [3]:
df = pd.read_csv('news.csv')

In [4]:
df

title  \
0                          You Can Smell Hillary’s Fear   
1     Watch The Exact Moment Paul Ryan Committed Pol...   
2           Kerry to go to Paris in gesture of sympathy   
3     Bernie supporters on Twitter erupt in anger ag...   
4      The Battle of New York: Why This Primary Matters   
...                                                 ...   
6330  State Department says it can't find emails fro...   
6331  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...   
6332  Anti-Trump Protesters Are Tools of the Oligarc...   
6333  In Ethiopia, Obama seeks progress on peace, se...   
6334  Jeb Bush Is Suddenly Attacking Trump. Here's W...   

                                                   text label  
0     Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1     Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2     U.S. Secretary of State John F. Kerry said Mon...  REAL  
3     — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4     It's primary day in New York and front-runners...  REAL  
...                                                 ...   ...  
6330  The State Department told the Republican Natio...  REAL  
6331  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...  FAKE  
6332  Anti-Trump Protesters Are Tools of the Oligarc...  FAKE  
6333  ADDIS ABABA, Ethiopia —President Obama convene...  REAL  
6334  Jeb Bush Is Suddenly Attacking Trump. Here's W...  REAL  

[6335 rows x 3 columns]

In [5]:
vectorizer = TfidfVectorizer()

In [6]:
features = vectorizer.fit_transform(df['text'])
# features2 = vectorizer.fit_transform(df['text'])
labels = df['label']

In [7]:
labels

0       FAKE
1       FAKE
2       REAL
3       FAKE
4       REAL
        ... 
6330    REAL
6331    FAKE
6332    FAKE
6333    REAL
6334    REAL
Name: label, Length: 6335, dtype: object

In [8]:
from sklearn.feature_extraction.text  import TfidfVectorizer
from sklearn.svm import SVC
from sklearn.metrics import classification_report, precision_score, recall_score

In [9]:
kf = KFold(n_splits=5)

In [10]:
for train, test in kf.split(features):
    model = SVC()
    model.fit(features[train], labels.iloc[train])
    predicts =  model.predict(features[test])
#     print('predictions:', predicts)
#     print('test:', labels[test])
#     print (classification_report(labels[test], predicts))

              precision    recall  f1-score   support

        FAKE       0.90      0.97      0.93       634
        REAL       0.97      0.90      0.93       633

    accuracy                           0.93      1267
   macro avg       0.93      0.93      0.93      1267
weighted avg       0.93      0.93      0.93      1267

              precision    recall  f1-score   support

        FAKE       0.91      0.97      0.94       663
        REAL       0.96      0.89      0.93       604

    accuracy                           0.93      1267
   macro avg       0.94      0.93      0.93      1267
weighted avg       0.93      0.93      0.93      1267

              precision    recall  f1-score   support

        FAKE       0.91      0.95      0.93       595
        REAL       0.96      0.91      0.93       672

    accuracy                           0.93      1267
   macro avg       0.93      0.93      0.93      1267
weighted avg       0.93      0.93      0.93      1267

              preci

After testing and training 6335 articles, we can trust that a model based on this dataset will have a precision between 91 and 97% and are satisfied with that score. 

In [13]:
def train_model():
    model = SVC()
    for train, test in kf.split(features):
        model.fit(features[train], labels.iloc[train])
        print ('hi')
    return model

model = train_model()

hi
hi
hi
hi
hi


In [14]:
from bs4 import BeautifulSoup
from bs4.element import Comment
import urllib.request


def delete_tag(element):
    if element.parent.name in ['style', 'script', 'head', 'title', 'meta', '[document]']: # returns false if text is a tag
        return False
    if isinstance(element, Comment): # returns false if text is a comment
        return False
    return True

def html_to_text(body):
    soup = BeautifulSoup(body, 'html.parser')  # goes through article and converts page into text type
    texts = soup.findAll(text=True)
    visible_texts = filter(delete_tag, texts)  
    return u" ".join(t.strip() for t in visible_texts)

# html = urllib.request.urlopen('https://www.washingtonpost.com/world/kerry-to-go-to-paris-in-gesture-of-sympathy/2015/01/12/c63fd031-20fc-4901-89e4-b637c9cf529e_story.html').read()
# text = html_to_text(html)

In [16]:
# if website blocks  from scraping, ask user for title of website

try:
    html = urllib.request.urlopen(input("Copy and Paste article url: ")).read()
    soup = BeautifulSoup(html, 'html.parser')
    title = soup.find('title')
    text = html_to_text(html)
except:
    print('Unable to access url of article.')
    title = input("Copy and Paste Title Name: ")
    text = input("Copy and Paste Article Text: ")
    
# print(text)

In [28]:
features = vectorizer.fit_transform([text])
model.predict(features)

ValueError: X.shape[1] = 359 should be equal to 67758, the number of features at training time

In [24]:
if model.predict(features) == ['FAKE']:
    print ('This article is fake. Find a better source.')
elif model.predict(features) == ['REAL']:
    print ('This article is real. You can trust this source.')
else:
    print ('There is an error.')

scipy.sparse.csr.csr_matrix

# At this moment in time, I am able to take a url and get the title and text from the site. Some sites block you from scraping their info, so if that is the case, the user can just copy and paste the info. However, I am having problems with actually predicting/ testing the user's article. 